### ISSUE TITLE
sample/color.rb error

### ISSUE THREAD

**Issue by dvarrui** [ruby/curses#69]
I'm using OpenSUSE 15.2 OS. And I'm running ruby 3.0,3 installed over rbenv and ruby 2.5.0 on real machine.
I had run all samples on my laptop without problems but when I run sample/colors.rb I get this error:

> ruby colors.rb
colors.rb:14:in `colors': colors() function is unimplemented on this machine (NotImplementedError)
	from colors.rb:14:in `<main>'

Could you give me any suggestion to fix it? Thanks!

**Comment by shugo** [comment]
Which version of underlying curses library do you use?

```
$ ruby -r curses -e 'p Curses::VERSION'
```

I guess the underlying library doesn't support colors().
Could you install ncurses (and a development package like ncurses-dev) before installing curses.gem?

**Comment by dvarrui** [comment]
[problem log file](https://github.com/dvarrui/proyectos-de-ejemplo/blob/master/ruby.ejemplos/03-graficos/curses/samples/problem-desc.md)

```
$ gem uninstall curses
$ sudo zypper install ncurses ncurses-devel
```

```
$ gem install curses

Fetching curses-1.4.3.gem
Building native extensions. This could take a while...
Successfully installed curses-1.4.3
Parsing documentation for curses-1.4.3
Installing ri documentation for curses-1.4.3
Done installing documentation for curses after 0 seconds
1 gem installed
```

```
$ ruby -r curses -e 'p Curses::VERSION'
"ncurses 6.1.20180317"
```

```
$ ./colors.rb
./colors.rb:14:in `colors': colors() function is unimplemented on this machine (NotImplementedError)
	from ./colors.rb:14:in `<main>'
```

**Comment by shugo** [comment]
Could you run extconf.rb and show the "have_var: checking for COLOR..." part of mkmf.log?

```
$ git clone https://github.com/ruby/curses
$ cd curses/ext/curses
$ ruby extconf.rb
```

The result of my machine is as follows:

```
have_var: checking for COLORS in ncursesw/curses.h... -------------------- yes

LD_LIBRARY_PATH=.:/home/shugo/.rbenv/versions/3.1.0/lib "gcc -I/home/shugo/.rbenv/versions/3.1.0/include/ruby-3.1.0/x86_64-linux
 -I/home/shugo/.rbenv/versions/3.1.0/include/ruby-3.1.0/ruby/backward -I/home/shugo/.rbenv/versions/3.1.0/include/ruby-3.1.0 -I.
 -I/home/shugo/.rbenv/versions/3.1.0/include    -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wdeprecated-declarations -Wduplicated-c
ond -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wwrite-strings -Wimplicit-fallthrou
gh=0 -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wn
o-overlength-strings -Wno-packed-bitfield-compat -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unuse
d-parameter -Wno-unused-value -Wsuggest-attribute=format -Wsuggest-attribute=noreturn -Wunused-variable -Wundef  -fPIC   -c conf
test.c"
checked program was:
/* begin */
 1: #include "ruby.h"
 2:
 3: #include <ncursesw/curses.h>
 4:
 5: /*top*/
 6: extern int t(void);
 7: int main(int argc, char **argv)
 8: {
 9:   if (argc > 1000000) {
10:     int (* volatile tp)(void)=(int (*)(void))&t;
11:     printf("%d", (*tp)());
12:   }
13:
14:   return !!argv[argc];
15: }
16: int t(void) { const volatile void *volatile p; p = &(&COLORS)[0]; return !p; }
/* end */
```

**Comment by dvarrui** [comment]
# mkmf.log

```
have_var: checking for ESCDELAY in ncursesw/curses.h... -------------------- no

"gcc -I/usr/include/ruby-2.5.0/x86_64-linux-gnu -I/usr/include/ruby-2.5.0/ruby/backward -I/usr/include/ruby-2.5.0 -I.    -fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -fno-strict-aliasing -fPIC   -c conftest.c"
conftest.c: In function ‘t’:
conftest.c:15:54: error: lvalue required as unary ‘&’ operand
 int t(void) { const volatile void *volatile p; p = &(&ESCDELAY)[0]; return !p; }
                                                      ^
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: #include <ncursesw/curses.h>
 4: 
 5: /*top*/
 6: extern int t(void);
 7: int main(int argc, char **argv)
 8: {
 9:   if (argc > 1000000) {
10:     printf("%p", &t);
11:   }
12: 
13:   return 0;
14: }
15: int t(void) { const volatile void *volatile p; p = &(&ESCDELAY)[0]; return !p; }
/* end */
```

**Comment by shugo** [comment]
It seems that it's not COLORS but ESCDELAY, but does the same error occur with COLORS?

If so, COLORS is not defined in your environment, but I don't know why.
If your terminal supports 256 colors, the following workaround may work:

```ruby
def Curses.colors
  return 256
end
```

**Comment by dvarrui** [comment]
```bash
> irb
```
```ruby
irb(main):001:0> require 'curses'
=> true
irb(main):002:0> def Curses.colors
irb(main):003:1> return 256
irb(main):004:1> end
=> :colors
irb(main):005:0> Curses.colors
=> 256
irb(main):006:0> quit
```

```bash
> ./colors.rb 
Traceback (most recent call last):
	1: from ./colors.rb:14:in `<main>'
./colors.rb:14:in `colors':colors() function is unimplemented on this machine (NotImplementedError)
```

```bash
>  more colors.rb 
```

```ruby
#!/usr/bin/env ruby

require "curses"
include Curses

begin
  init_screen

  if !Curses.has_colors?
    addstr "This Terminal does not support colors!"
  else
    start_color

    addstr "This Terminal supports #{Curses.colors} colors.\n"

    Curses.colors.times { |i|
      Curses.init_pair(i, i, 0)
      attrset(color_pair(i))
      addstr("#{i.to_s.rjust(3)} ")
      addstr("\n") if i == 15 || (i > 16 && (i - 15) % 36 == 0)
    }
  end

  getch

ensure
  close_screen
end
```

System OS: OpenSUSE 15.3
System ruby 2.5
rbenv 1.2.0-6
rbenv version => 3.0.0

**Comment by shugo** [comment]
Add the following code to colors.rb:

```ruby
def Curses.colors
  return 256
end
```

**Comment by dvarrui** [comment]
I added this 2 methods because the code call both on line 16 and 18.
```
#!/usr/bin/env ruby

require "curses"
include Curses

def colors
  return 256
end

def Curses.colors
  return 256
end

...
```

And works fine! ;-) 
Thanks!
Will you change colors.rb sample?

**Comment by dvarrui** [comment]
I would like to say something more....

I proposs 2 methods... to solve colors.rb errors but I think wiht only 1 It could be nice... 
I explain this better...

If this method is defined into Curses module:
```
def Curses.colors
  256
end
```

Then when colors.rb does "include Curses"... then "colors" method and "Curses.colors" methods will be both defined (really are only one method) into colors.rb sample.

What do you  think?

Thanks! :-)

![Teide from Tenerife](https://aws-tiqets-cdn.imgix.net/images/content/b5a51e313581409caa57c33c1f062474.jpg?auto=format&fit=crop&h=800&ixlib=python-3.2.1&q=70&w=800&s=8ff5cf11259158fb3258d55830b33b15)

**Comment by shugo** [comment]
Defining Curses.colors is just a workaround, and it's not good to add it in sample/color.rb because supported colors depends on terminals.

COLORS may be a macro on your environment.

Does the following patch work?

```
diff --git a/ext/curses/curses.c b/ext/curses/curses.c
index 6b4c68e..8f3c4cc 100644
--- a/ext/curses/curses.c
+++ b/ext/curses/curses.c
@@ -1366,7 +1366,7 @@ curses_can_change_color(VALUE obj)
     return can_change_color() ? Qtrue : Qfalse;
 }
 
-#if defined(HAVE_COLORS)
+#if defined(HAVE_COLORS) || defined(COLORS)
 /*
  * Document-method: Curses.color
  *
```

**Comment by dvarrui** [comment]
I would like to say something more....

I proposs 2 methods... to solve colors.rb errors but I think wiht only 1 It could be nice... 
I explain this better...

If this method is defined into Curses module:
```
def Curses.colors
  256
end
```

Then when colors.rb does "include Curses"... then "colors" method and "Curses.colors" methods will be both defined (really are only one method) into colors.rb sample.

What do you  think?

Thanks! :-)

![Teide from Tenerife](https://aws-tiqets-cdn.imgix.net/images/content/b5a51e313581409caa57c33c1f062474.jpg?auto=format&fit=crop&h=800&ixlib=python-3.2.1&q=70&w=800&s=8ff5cf11259158fb3258d55830b33b15)

**Comment by dvarrui** [comment]
```bash
dalmata:/ # ruby -v
ruby 2.5.9p229 (2021-04-05 revision 67939) [x86_64-linux-gnu]
```

I'm sorry! I can't or I don't know how to run `diff --git a/ext/curses/curses.c b/ext/curses/curses.c`... I don't know where a and b folders are.

```bash
❯ v
total 28
-rw-r--r-- 1 david users 1282 feb  1 23:33 BSDL
-rw-r--r-- 1 david users 2504 feb  1 23:33 COPYING
-rw-r--r-- 1 david users  846 feb  1 23:33 curses.gemspec
drwxr-xr-x 3 david users   20 feb  1 23:33 ext/
-rw-r--r-- 1 david users   69 feb  1 23:33 Gemfile
-rw-r--r-- 1 david users 3964 feb  1 23:33 History.md
drwxr-xr-x 2 david users   23 feb  1 23:33 lib/
-rw-r--r-- 1 david users   52 feb  1 23:33 Rakefile
-rw-r--r-- 1 david users 2385 feb  1 23:33 README.md
drwxr-xr-x 2 david users  188 feb  1 23:34 sample/
drwxr-xr-x 3 david users   22 feb  1 23:33 vendor/

 ~/proy/repos/ruby.d/external/curses  master 
```

I tried to do it (diif I mean) with this...

```bash
dalmata:/ # find / -name curses.c
/home/david/proy/repos/ruby.d/external/curses/ext/curses/curses.c
/home/david/.rbenv/versions/3.0.3/lib64/ruby/gems/3.0.0/gems/curses-1.4.3/ext/curses/curses.c
/home/david/.rbenv/versions/2.7.5/lib64/ruby/gems/2.7.0/gems/curses-1.4.3/ext/curses/curses.c
find: ‘/run/user/1000/doc’: Permiso denegado
find: ‘/run/user/1000/gvfs’: Permiso denegado
/usr/lib64/ruby/gems/2.5.0/gems/curses-1.4.2/ext/curses/curses.c
```

```bash
dalmata:/ # diff /usr/lib64/ruby/gems/2.5.0/gems/curses-1.4.2/ext/curses/curses.c /home/david/proy/repos/ruby.d/external/curses/ext/curses/curses.c 
623a624,637
>  * Document-method: Curses.flushinp
>  *
>  * The flushinp routine throws away any typeahead that has been
>  * typed by the user and has not yet been read by the program.
>  */
> static VALUE
> curses_flushinp(VALUE obj)
> {
>     curses_stdscr();
>     flushinp();
>     return Qnil;
> }
> 
> /*
3621a3636,3750
>  * Document-method: Curses::Menu#mark=
>  *
>  * call-seq:
>  *   mark=(str)
>  *
>  * Set the mark string to distinguish the selected items
>  */
> static VALUE
> menu_set_mark(VALUE obj, VALUE mark)
> {
>     struct menudata *menup;
> 
>     GetMENU(obj, menup);
>     set_menu_mark(menup->menu, StringValueCStr(mark));
> 
>     return obj;
> }
> 
> /*
>  * Document-method: Curses::Menu#mark
>  *
>  * call-seq:
>  *   mark
>  *
>  * Get the Menu's mark string
>  */
> static VALUE
> menu_get_mark(VALUE obj)
> {
>     struct menudata *menup;
>     const char *mark;
> 
>     GetMENU(obj, menup);
>     mark = menu_mark(menup->menu);
> 
>     return rb_external_str_new_with_enc(mark, strlen(mark), terminal_encoding);
> }
> 
> /*
>  * Document-method: Curses::Menu#fore=
>  *
>  * call-seq:
>  *   fore=(attr)
>  *
>  * Sets the foreground attribute of menu.
>  * This is the highlight used for selected menu items.
>  */
> static VALUE
> menu_set_fore(VALUE obj, VALUE attr)
> {
>     struct menudata *menup;
> 
>     GetMENU(obj, menup);
>     set_menu_fore(menup->menu, NUM2CHTYPE(attr));
> 
>     return attr;
> }
> 
> /*
>  * Document-method: Curses::Menu#fore
>  *
>  * call-seq:
>  *   fore
>  *
>  * Sets the foreground attribute of menu.
>  * This is the highlight used for selected menu items.
>  */
> static VALUE
> menu_get_fore(VALUE obj, VALUE attr)
> {
>     struct menudata *menup;
> 
>     GetMENU(obj, menup);
> 
>     return CHTYPE2NUM(menu_fore(menup->menu));
> }
> 
> /*
>  * Document-method: Curses::Menu#set_back
>  *
>  * call-seq:
>  *   set_back(attr)
>  *
>  * Get the background attribute of menu.
>  */
> static VALUE
> menu_set_back(VALUE obj, VALUE attr)
> {
>     struct menudata *menup;
> 
>     GetMENU(obj, menup);
>     CHTYPE2NUM(set_menu_back(menup->menu, NUM2CHTYPE(attr)));
> 
>     return attr;
> }
> 
> /*
>  * Document-method: Curses::Menu#back
>  *
>  * call-seq:
>  *   back
>  *
>  * Get the background attribute of menu.
>  */
> static VALUE
> menu_get_back(VALUE obj, VALUE attr)
> {
>     struct menudata *menup;
> 
>     GetMENU(obj, menup);
> 
>     return CHTYPE2NUM(menu_back(menup->menu));
> }
> 
> /*
4755a4885
>     rb_define_module_function(mCurses, "flushinp", curses_flushinp, 0);
4990a5121,5126
>     rb_define_method(cMenu, "mark=", menu_set_mark, 1);
>     rb_define_method(cMenu, "mark", menu_get_mark, 0);
>     rb_define_method(cMenu, "fore=", menu_set_fore, 1);
>     rb_define_method(cMenu, "fore", menu_get_fore, 0);
>     rb_define_method(cMenu, "back=", menu_set_back, 1);
>     rb_define_method(cMenu, "back", menu_get_back, 0);
dalmata:/ # 
```

**Comment by shugo** [comment]
I've released curses-1.4.4.beta.1 applied the patch.
Please try it using the --pre option of the gem command:

```
$ gem  install curses --pre
Building native extensions. This could take a while...
Successfully installed curses-1.4.4.beta.1
Parsing documentation for curses-1.4.4.beta.1
Done installing documentation for curses after 0 seconds
1 gem installed
```

**Comment by shugo** [comment]
Thanks for your confirmation.  I'll release curses-1.4.4 later.

99464d6 is the fix which define Curses.colors even if COLORS is not a variable but a macro.
